How many late days are you using for this assignment? 

In [109]:
# Imports and pip installations (if needed)

import pandas as pd
from sklearn.preprocessing import StandardScaler


# Part 1: Load the dataset

In [110]:
# Load the given datasets

categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")
numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data

display(categorical)
display(numerical)


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: Each entry/data gives a patients' (identified by id) medical records. The differentg features and labels identify specific medical categories that are evaluated and stores. The different features are age, blood pressure (bp), Specific Gravity (sg), Albumin(al), Sugar(su), Red Blood Cells(rbc), Pus Cell (pc), Pus Cell clumps(pcc), Bacteria(ba), Blood Glucose Random(bgr), Blood Urea (bu), Serum Creatinine (sc), Sodium(sod), Potassium(pot), Hemoglobin(hemo), Packed Cell Volume, White Blood Cell Count(wc), Red Blood Cell Count(rc), Hypertension(htn), Diabetes Mellitus(dm), Coronary Artery Disease(cad), Appetite(appet), Pedal Edema(pe), Anemia(ane), Class. 3 Categorical Examples: Specific Gravity, Albumin, Sugar. 3 Numerical Examples: Age, Blood Pressure, Blood Glucose Random.  

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [111]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows

duplicates_num_sum = numerical.duplicated(subset=['unique_id'], keep='first').sum()
print("Total Number of Duplicate Rows: ")
print(duplicates_num_sum)

# Drop these duplicate rows

numerical_filtered = numerical.drop_duplicates(subset=['unique_id'], keep='first')

# Repeat the same for categorical dataset. Print the duplicate rows and drop them

categorical_num_sum = categorical.duplicated(subset=['unique_id'], keep='first').sum()
print("Total Number of Duplicate Rows: ")
print(categorical_num_sum)

# Drop these duplicate rows

categorical_filtered = categorical.drop_duplicates(subset=['unique_id'], keep='first')


Total Number of Duplicate Rows: 
30
Total Number of Duplicate Rows: 
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, not all unique IDs are present in both datasets; some IDs exist only in the numerical dataset, while others exist only in the categorical dataset. The missing rows in the combined table have 'NaN' values in the columns corresponding to the dataset they are missing from.

In [112]:
# Merge the two given numerical and categorical datasets based on their unique_ID.

merged_dataset = pd.merge(numerical_filtered, categorical_filtered, on='unique_id',how='outer')

#Print the combined dataset

display(merged_dataset)


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,103570,24.0,80.0,125.0,NaN,NaN,136.0,3.5,15.4,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,104675,65.0,80.0,215.0,133.0,2.5,NaN,NaN,13.2,41.0,...,normal,present,notpresent,no,yes,no,good,no,no,ckd
3,105923,90.0,90.0,139.0,89.0,3.0,140.0,4.1,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
4,107297,60.0,70.0,NaN,NaN,NaN,NaN,NaN,16.4,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,...,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd
396,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,996412,70.0,60.0,120.0,40.0,0.5,140.0,4.6,16.0,43.0,...,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
398,997629,35.0,NaN,99.0,30.0,0.5,135.0,4.9,15.4,48.0,...,normal,NaN,NaN,no,no,no,good,no,no,notckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [113]:
# Calculate the percentage of rows that contain atleast one missing value

total_entries = len(merged_dataset)
missing_entries = merged_dataset.isnull().any(axis=1).sum()
percentage = (missing_entries / total_entries) * 100

# Print %

print(f"Percentage of Entries the contain One Missing Value: {percentage:.2f}%")

# Drop these rows from the dataset

merged_filtered = merged_dataset.dropna()

# Print the Dataset

display(merged_filtered)

Percentage of Entries the contain One Missing Value: 61.75%


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
6,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
8,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
390,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,...,abnormal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
391,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
394,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,...,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd


## Part 3.4: Sort the dataset according to the Labels

In [114]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting

dataset_sorted = merged_filtered.sort_values(by='Target').reset_index(drop=True)

# Print the dataset

display(dataset_sorted)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,...,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,...,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [115]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dataset_dummy = pd.get_dummies(dataset_sorted, columns=categorical_columns, dtype=int)

# Print the dataset

display(dataset_dummy)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,1,0,0,1,0,1,0,1,1,0
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,...,1,0,0,1,0,1,0,1,1,0
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,0,1,0,1,0,1,0,1,1,0
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,...,0,1,1,0,1,0,1,0,1,0
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,...,1,0,0,1,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,1,0,1,0,1,0,1,0,0,1
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,...,1,0,1,0,1,0,1,0,0,1
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,...,1,0,1,0,1,0,1,0,0,1
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,...,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: For a binary variable, using a single column (0 or 1) is equivalent to one-hot encoding in terms of model interpretability. On the flip side, If a categorical variable has more than two values (like "type of condiment" with 5 options), using a single column would lose information because it is no longer a yes or no.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [116]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for col in numerical_columns:
    if col in dataset_dummy.columns:
        mean = dataset_dummy[col].mean()
        std_dev = dataset_dummy[col].std()
        lower_bound = mean - (3 * std_dev)
        upper_bound = mean + (3 * std_dev)
        dataset_dummy = dataset_dummy[(dataset_dummy[col] >= lower_bound) & (dataset_dummy[col] <= upper_bound)]

# Print the dataset
display(dataset_dummy)


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,0,1,0,1,0,1,0,1,1,0
5,397388,63.0,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,...,0,1,1,0,1,0,1,0,1,0
7,438182,73.0,80.0,253.0,142.0,4.6,138.0,5.8,10.5,33.0,...,0,1,1,0,1,0,1,0,1,0
8,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,1,0,1,0,1,0,0,1,1,0
9,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,1,0,1,0,1,0,1,0,0,1
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,...,1,0,1,0,1,0,1,0,0,1
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,...,1,0,1,0,1,0,1,0,0,1
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,...,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [117]:
# Normalize the all Numerical Attributes in the dataset.

scaler = StandardScaler()
dataset_dummy[numerical_columns] = scaler.fit_transform(dataset_dummy[numerical_columns])

# Print the dataset

display(dataset_dummy)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,0.810190,0.730553,1.146438,4.014305,1.690051,-0.959702,1.373024,-2.913221,-2.905503,...,0,1,0,1,0,1,0,1,1,0
5,397388,0.936367,2.776102,3.520558,-0.225995,1.218114,0.448743,-1.303189,-0.607180,-0.590673,...,0,1,1,0,1,0,1,0,1,0
7,438182,1.567253,0.730553,2.921481,3.789156,2.161987,-0.431535,2.317570,-1.694935,-1.603411,...,0,1,1,0,1,0,1,0,1,0
8,474407,-1.713353,1.753327,-0.317973,-0.038371,0.206821,-2.720258,-1.303189,-2.652160,-3.050180,...,1,0,1,0,1,0,0,1,1,0
9,475200,0.999456,2.776102,0.924557,0.486976,3.914895,-0.079424,0.428478,-2.826200,-2.616150,...,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,-1.776441,-0.292221,0.037036,0.111728,-0.265116,-0.959702,-0.830916,0.088983,-0.011966,...,1,0,1,0,1,0,1,0,0,1
149,203466,-1.650264,-1.314996,-0.539853,-0.863916,-0.130277,-0.431535,-0.043795,-0.389629,-0.301320,...,1,0,1,0,1,0,1,0,0,1
150,482105,-0.262316,-1.314996,-0.584229,0.186777,-0.602214,-0.431535,-0.201219,0.263024,0.856096,...,1,0,1,0,1,0,1,0,0,1
151,228849,0.620924,0.730553,-0.473289,0.336877,-0.130277,-0.079424,-1.303189,-0.172078,0.856096,...,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The 'unique_id' cloumn is not appropriate and should be removed as it is simply an identifer for a patient. It does not provide any medical information to assist in any sort of modeling and predictions and should therefore be removed.

In [118]:
#Remove that column

cleaned_dataset = dataset_dummy.drop(columns=['unique_id'])

# Print the dataset

print(cleaned_dataset)

          age        bp       bgr        bu        sc       sod       pot  \
2    0.810190  0.730553  1.146438  4.014305  1.690051 -0.959702  1.373024   
5    0.936367  2.776102  3.520558 -0.225995  1.218114  0.448743 -1.303189   
7    1.567253  0.730553  2.921481  3.789156  2.161987 -0.431535  2.317570   
8   -1.713353  1.753327 -0.317973 -0.038371  0.206821 -2.720258 -1.303189   
9    0.999456  2.776102  0.924557  0.486976  3.914895 -0.079424  0.428478   
..        ...       ...       ...       ...       ...       ...       ...   
148 -1.776441 -0.292221  0.037036  0.111728 -0.265116 -0.959702 -0.830916   
149 -1.650264 -1.314996 -0.539853 -0.863916 -0.130277 -0.431535 -0.043795   
150 -0.262316 -1.314996 -0.584229  0.186777 -0.602214 -0.431535 -0.201219   
151  0.620924  0.730553 -0.473289  0.336877 -0.130277 -0.079424 -1.303189   
152  0.557836 -1.314996 -0.362349  0.299352 -0.130277  1.681133  0.585903   

         hemo       pcv      wbcc  ...  cad_no  cad_yes  appet_good  \
2   

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [119]:
# Export the dataframe to a new csv file

cleaned_dataset.to_csv('cleaned_dataframe.csv', index=False)

# Export the dataframe to a new json file

cleaned_dataset.to_json('cleaned_dataframe.json', orient='records', lines=True)


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [120]:
print(cleaned_dataset.head(15))

         age        bp       bgr        bu        sc       sod       pot  \
2   0.810190  0.730553  1.146438  4.014305  1.690051 -0.959702  1.373024   
5   0.936367  2.776102  3.520558 -0.225995  1.218114  0.448743 -1.303189   
7   1.567253  0.730553  2.921481  3.789156  2.161987 -0.431535  2.317570   
8  -1.713353  1.753327 -0.317973 -0.038371  0.206821 -2.720258 -1.303189   
9   0.999456  2.776102  0.924557  0.486976  3.914895 -0.079424  0.428478   
10  0.747101 -1.314996  3.698062 -0.188470  0.206821 -1.839980 -2.090311   
12  0.494747  1.753327  0.170164  2.475789  3.577798 -1.663925  0.743327   
13  1.314898 -0.292221  2.056147  2.063016  3.308119 -3.600537 -0.673492   
15 -0.136138 -1.314996  0.924557  1.912917  1.285533  0.096632 -0.516068   
18  0.873279 -0.292221  0.014848  0.036678  0.206821 -0.783646  0.585903   
19  0.431659  0.730553  2.056147  1.199946  1.690051 -0.607591  0.900751   
20  0.999456 -1.314996  2.610848  0.637075  1.959729 -0.607591  1.687873   
22  1.441076

### Paste the markdown table here
|   age     |     bp     |     bgr     |      bu     |      sc     |     sod     |     pot     |    hemo    |     pcv    |    wbcc    | cad_no | cad_yes | appet_good |
|-----------|------------|-------------|-------------|-------------|-------------|-------------|------------|------------|------------|--------|---------|------------|
| 0.810190  | 0.730553   | 1.146438    | 4.014305    | 1.690051    | -0.959702   | 1.373024    | -2.913221  | -2.905503  | 0.519658   | 0      | 1       | 0          |
| 0.936367  | 2.776102   | 3.520558    | -0.225995   | 1.218114    | 0.448743    | -1.303189   | -0.607180  | -0.590673  | 0.785483   | 0      | 1       | 1          |
| 1.567253  | 0.730553   | 2.921481    | 3.789156    | 2.161987    | -0.431535   | 2.317570    | -1.694935  | -1.603411  | -0.366426  | 0      | 1       | 1          |
| -1.713353 | 1.753327   | -0.317973   | -0.038371   | 0.206821    | -2.720258   | -1.303189   | -2.652160  | -3.050180  | 1.937392   | 1      | 0       | 1          |
| 0.999456  | 2.776102   | 0.924557    | 0.486976    | 3.914895    | -0.079424   | 0.428478    | -2.826200  | -2.616150  | -0.233513  | 1      | 0       | 1          |
| 0.747101  | -1.314996  | 3.698062    | -0.188470   | 0.206821    | -1.839980   | -2.090311   | -2.826200  | -2.760826  | 3.177909   | 1      | 0       | 0          |
| 0.494747  | 1.753327   | 0.170164    | 2.475789    | 3.577798    | -1.663925   | 0.743327    | -2.304078  | -2.182119  | -0.720859  | 1      | 0       | 1          |
| 1.314898  | -0.292221  | 2.056147    | 2.063016    | 3.308119    | -3.600537   | -0.673492   | -1.621407  | -1.494122  | -1.672854  | 1      | 0       | 1          |
| -0.136138 | -1.314996  | 0.924557    | 1.912917    | 1.285533    | 0.096632    | -0.516068   | -1.151843  | -1.207197  | 0.519658   | 0      | 1       | 0          |
| 0.873279  | -0.292221  | 0.014848    | 0.036678    | 0.206821    | -0.783646   | 0.585903    | -1.462310  | -1.437276  | 1.255014   | 0      | 1       | 0          |
| 0.431659  | 0.730553   | 2.056147    | 1.199946    | 1.690051    | -0.607591   | 0.900751    | -2.113778  | -2.344025  | -0.738690  | 0      | 1       | 0          |
| 0.999456  | -1.314996  | 2.610848    | 0.637075    | 1.959729    | -0.607591   | 1.687873    | -0.500778  | -0.457117  | -0.938083  | 1      | 0       | 0          |
| 1.441076  | -1.314996  | -0.073905   | 3.151235    | 2.633924    | -0.783646   | 0.900751    | -0.198452  | 1.436751   | -2.234080  | 1      | 0       | 1          |
| 1.441076  | -0.292221  | 2.167088    | 1.537669    | 1.487792    | -1.311813   | 0.113630    | -1.581150  | 0.284328   | 0.084494   | 0      | 1       | 0          |
| 0.684013  | 0.730553   | 4.030883    | -0.225995   | -0.062857   | -3.248425   | -1.303189   | -1.664310  | -1.839980  | 1.002554   | 1      | 0       | 1          |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

A good way to mitigate data conversion hallucinations from LLM conversions would be to always validate outputs of the LLM and also simply just use code-based conversions instead of relying on direct LLM conversions.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [121]:
# Code to rename all the columns in the dataset

rename_dict = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc_abnormal': 'Red Blood Cell Abnormal',
    'rbc_normal': 'Red Blood Cell Normal',
    'pc_abnormal': 'Pus Cell Abnormal',
    'pc_normal': 'Pus Cell Normal',
    'pcc_notpresent': 'Pus Cell Cast Not Present',
    'pcc_present': 'Pus Cell Cast Present',
    'ba_notpresent': 'Bacteria Not Present',
    'ba_present': 'Bacteria Present',
    'htn_no': 'Hypertension No',
    'htn_yes': 'Hypertension Yes',
    'dm_no': 'Diabetes Mellitus No',
    'dm_yes': 'Diabetes Mellitus Yes',
    'cad_no': 'Coronary Artery Disease No',
    'cad_yes': 'Coronary Artery Disease Yes',
    'appet_good': 'Good Appetite',
    'appet_poor': 'Poor Appetite',
    'pe_no': 'Pedal Edema No',
    'pe_yes': 'Pedal Edema Yes',
    'ane_no': 'Anemia No',
    'ane_yes': 'Anemia Yes',
}

cleaned_dataset.rename(columns=rename_dict, inplace=True)

# To verify the new column names
display(cleaned_dataset)

,Age,Blood Pressure,Blood Glucose Random,Blood Urea,Serum Creatinine,Sodium,Potassium,Hemoglobin,Packed Cell Volume,White Blood Cell Count,...,Coronary Artery Disease No,Coronary Artery Disease Yes,Good Appetite,Poor Appetite,Pedal Edema No,Pedal Edema Yes,Anemia No,Anemia Yes,Target_ckd,Target_notckd
2,0.810190,0.730553,1.146438,4.014305,1.690051,-0.959702,1.373024,-2.913221,-2.905503,0.519658,...,0,1,0,1,0,1,0,1,1,0
5,0.936367,2.776102,3.520558,-0.225995,1.218114,0.448743,-1.303189,-0.607180,-0.590673,0.785483,...,0,1,1,0,1,0,1,0,1,0
7,1.567253,0.730553,2.921481,3.789156,2.161987,-0.431535,2.317570,-1.694935,-1.603411,-0.366426,...,0,1,1,0,1,0,1,0,1,0
8,-1.713353,1.753327,-0.317973,-0.038371,0.206821,-2.720258,-1.303189,-2.652160,-3.050180,1.937392,...,1,0,1,0,1,0,0,1,1,0
9,0.999456,2.776102,0.924557,0.486976,3.914895,-0.079424,0.428478,-2.826200,-2.616150,-0.233513,...,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,-1.776441,-0.292221,0.037036,0.111728,-0.265116,-0.959702,-0.830916,0.088983,-0.011966,-1.119597,...,1,0,1,0,1,0,1,0,0,1
149,-1.650264,-1.314996,-0.539853,-0.863916,-0.130277,-0.431535,-0.043795,-0.389629,-0.301320,-0.056296,...,1,0,1,0,1,0,1,0,0,1
150,-0.262316,-1.314996,-0.584229,0.186777,-0.602214,-0.431535,-0.201219,0.263024,0.856096,-0.144905,...,1,0,1,0,1,0,1,0,0,1
151,0.620924,0.730553,-0.473289,0.336877,-0.130277,-0.079424,-1.303189,-0.172078,0.856096,-0.587946,...,1,0,1,0,1,0,1,0,0,1


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [122]:
# Converted SQL to Pandas code

result = cleaned_dataset.groupby('Target_ckd').size().reset_index(name='count')

display(result)

,Target_ckd,count
0,0,111
1,1,22


*Describe what the above code does here* This code counts how many times in the dataset each possible value of 'Target_ckd' occurred in the dataset. the value 0 occurred 111 times where 1 occurred 22 times.